### PR TITLE
Implement electric characters

### LINKED ISSUE #1133: Implement electric comma, semicolon, etc
`haskell-indentation` reacts to a couple of characters, most notably comma and semicolonbut also other chars.

Implement function that can be used in `post-command-hook` that will:
- [ ] check if there is exactly one character on the line
- [ ] check if there is only one indentation position returned from `haskell-indentation-find-indentations`
- [ ] do the indentation if both are true
- [ ] connect this function to `,;` and `)}]`

More info:
http://www.gnu.org/software/emacs/manual/html_node/ccmode/Electric-Keys.html

### LINKED ISSUE #1294: Unused lexical variables (Lexical Binding)
*Is this support for Emacs 24.5 version?*   
*Has small problem with some varable.*  

```
Compiling file /haskell-mode-20160419.752/haskell-indentation.el at Thu Apr 21 20:58:35 2016
haskell-indentation.el:657:1:Warning: Unused lexical variable `starter-indent-inside'
haskell-indentation.el:1024:1:Warning: Unused lexical variable `starter-line'

Compiling file /haskell-mode-20160419.752/haskell-load.el at Thu Apr 21 20:58:36 2016
haskell-load.el:93:1:Warning: Unused lexical argument `e'

Compiling file /haskell-mode-20160419.752/haskell-process.el at Thu Apr 21 20:58:36 2016
haskell-process.el:488:1:Warning: Unused lexical argument `e'
```

### PR DESCRIPTION
**Pull request by fice-t**
This does it like in `cc-mode'.

Also, I removed a few unnecessary things in the minor mode definition and removed an unused variable (Fixes #1294).

Closes #1133

### CODE REVIEW

File: haskell-indentation.el [hunk 1]
```diff
@@ -323,7 +340,42 @@ fixes up only indentation."
...
+2) There is only one possible indentation position.
+3) The variable `haskell-use-electric-keys' is non-nil.
+4) The point is not in a comment, string, or quasiquote."
+  (interactive "*p")
+  (let* ((col (haskell-indentation-current-indentation))
+         (at-indent? (= col (current-column)))
+         ind)
+    (self-insert-command arg)
+    (when (and at-indent?
+               haskell-indentation-electric-flag
```
Comment (gracjan) [inline haskell-indentation.el]: Change this to:

```haskell
(or (equal 'always haskell-indentation-electric-flag)
    (bound-and-true-p 'electric-indent-mode))
```

Then the mode changing hook will not be necessary.

Comment (fice-t) [inline haskell-indentation.el]: Thanks, I don't know why I didn't think of that. At that point `haskell-indentation-electric-flag` might as well just be used as a boolean, so I changed it accordingly.

File: haskell-indentation.el [hunk 2]
```diff
@@ -323,7 +331,31 @@ fixes up only indentation."
...
+  "Call `self-insert-command' to insert the character typed ARG times
+and indent when all of the following are true:
+
+1) The character is the first non-whitespace character on the line.
+2) There is only one possible indentation position.
+3) The variable `electric-indent-mode' or `haskell-indentation-electric-flag'
+   is non-nil.
+4) The point is not in a comment, string, or quasiquote."
+  (interactive "*p")
+  (let* ((col (haskell-indentation-current-indentation))
```
Comment (gracjan) [inline haskell-indentation.el]: Also: please check the `(or haskell-indentation-electric-flag electric-indent-mode)` in the beginning of the function. The point is that `haskell-indentation-current-indentation` may possibly be buggy, take lots of time or error-out. Therefore it should be invoked only after flag is enabled.

File: haskell-indentation.el [hunk 3]
```diff
@@ -87,18 +100,13 @@
 It rebinds RET, DEL and BACKSPACE, so that indentations can be
 set and deleted as if they were real tabs."
   :keymap haskell-indentation-mode-map
-  (kill-local-variable 'indent-line-function)
```
Comment (gracjan) [inline haskell-indentation.el]: This `kill-local-variable`s should stay here. Reason: minor modes may be switched on and off, this code path is for switching off (although it is executed always).

Comment (fice-t) [inline haskell-indentation.el]: Right, I was thinking that someone would always want either haskell-indent or haskell-indentation. I guess someone might want neither.

**Comment by gracjan** [comment]
Code looks good, I plan to run it locally for a couple of days to see how it 'feels'.

**Comment by gracjan** [comment]
We should have a nice documentation page in manual. @fice-t, can you write it in `doc/haskell-mode.texi`? It is pretty easy.

**Comment by fice-t** [comment]
Should that do for documentation?

**Pull request merged**